File: modules/guides/examples/search/autocomplete-rest-endpoint.js
const Couchbase = require("couchbase");
const Express = require("express");
const Cors = require("cors");
 
var app = Express();
var cluster = new Couchbase.Cluster("${CB_HOSTNAME}");
cluster.authenticate("${CB_USERNAME", "${CB_PASSWORD}");
var bucket = cluster.openBucket("${CB_BUCKET}");
 
app.use(Cors());
 
app.get("/search", (request, response) => { });
 
app.listen(3000, () => {
    console.log("Listening at :3000...");
});

app.get("/search", (request, response) => {
    var SearchQuery = Couchbase.SearchQuery;
    var query = SearchQuery.new("idx-music", SearchQuery.match(request.query.query).fuzziness(1));
    query.fields(["title", "artist"]);
    bucket.query(query, function(error, result, meta) {
        if(error) {
            return response.status(500).send({ message: error.message });
        }
        result = result.map(item => {
            return {
                id: item.id,
                score: item.score,
                title: item.fields.title,
                artist: item.fields.artist
            };
        });
        response.send(result);
    });
});

File: modules/guides/examples/search/autocomplete-rest-endpoint.go
package main
 
import (
 "encoding/json"
 "net/http"
 
 "github.com/gorilla/handlers"
 "github.com/gorilla/mux"
 gocb "gopkg.in/couchbase/gocb.v1"
 "gopkg.in/couchbase/gocb.v1/cbft"
)
 
type Song struct {
 Id     string  `json:"id"`
 Score  float64 `json:"score"`
 Artist string  `json:"artist"`
 Title  string  `json:"title"`
}
 
var bucket *gocb.Bucket
 
func SearchEndpoint(response http.ResponseWriter, request *http.Request) { }
 
func main() {
 cluster, _ := gocb.Connect("${CB_HOSTNAME}")
 cluster.Authenticate(gocb.PasswordAuthenticator{Username: "${CB_USERNAME}", Password: "${CB_PASSWORD}"})
 bucket, _ = cluster.OpenBucket("example", "")
 router := mux.NewRouter()
 router.HandleFunc("/search", SearchEndpoint).Methods("GET")
 http.ListenAndServe(":12345", handlers.CORS(handlers.AllowedHeaders([]string{"X-Requested-With", "Content-Type", "Authorization"}), handlers.AllowedMethods([]string{"GET", "POST", "PUT", "HEAD", "OPTIONS"}), handlers.AllowedOrigins([]string{"*"}))(router))
}

response.Header().Set("content-type", "application/json")
 params := request.URL.Query()
 query := gocb.NewSearchQuery("idx-music", cbft.NewMatchQuery(params.Get("query")))
 query.Fields("artist", "title")
 result, _ := bucket.ExecuteSearchQuery(query)
 var hits []Song
 for _, hit := range result.Hits() {
 hits = append(hits, Song{
 Id:     hit.Id,
 Score:  hit.Score,
 Artist: hit.Fields["artist"],
 Title:  hit.Fields["title"],
 })
 }
 json.NewEncoder(response).Encode(hits)
}